posts/2023-XX-YY-boa-release-17.md

```diff
@@ -0,0 +1,215 @@
+let dyn_loader: &dyn ModuleLoader = loader;
+let mut context = &mut Context::builder().module_loader(dyn_loader).build().unwrap();
+
+let source = Source::from_bytes("1 + 3");
+
+let module = Module::parse(source, None, context).unwrap();
+
+loader.insert(Path::new("main.mjs").to_path_buf(), module.clone());
+
+let promise = module.load_link_evaluate(context).unwrap();
+
+context.run_jobs();
+
+assert_eq!(promise.state().unwrap(), PromiseState::Fulfilled(JsValue::undefined()));
+```
+
+For a more extensive, descriptive example that uses a real directory, you can check out [**boa_examples**][mod_example].
+
+[modules]: https://tc39.es/ecma262/#sec-modules
+[mod_loader]: https://boajs.dev/boa/doc/boa_engine/module/trait.ModuleLoader.html
+[mod_example]: https://github.com/boa-dev/boa/blob/main/boa_examples/src/bin/modules.rs
+
+### New APIs
+
+We have added new built-in object wrappers, such as [`JsPromise`][js_promise], [`JsRegExp`][js_regexp], [`JsGenerator`],
+[`JsDate`][js_date] and [`JsDataView`][js_data_view]. You can check all of them [here][wrappers].
+
+Then, we want to present you a new trait that we have developed to make it easier for you to interoperate between Rust
+and JavaScript: [`TryFromJs`][try_from_js]. All built-ins and Rust basic types that exist in JavaScript implement this
+trait, and it adds a new static method to them that allows you to convert a [`JsValue`][js_value] into a Rust structure.
+You can also convert any `JsValue` to a `TryFromJs` Rust type with [`JsValue::try_js_into()`][try_js_into] function.
+
+```rust
+let js_str = r#"
+let x = /[a-z0-9]@[a-z0-9]/;
+x;
+"#;
+
+let js = Source::from_bytes(js_str);
+let mut context = Context::default();
+let res = context.eval(js).unwrap();
+
+let rs_regexp: JsRegExp = res.try_js_into(context).unwrap();
+
+let test_result = rs_regexp.test("hello@domain", context)?;
+assert!(test_result);
+```
+
+Moreover, you can derive `TryFromJs` for any Rust structure, and in the case that you want to manually convert some of the struct attributes, you can override it:
+
+```rust
+/// Converts the value lossly
+fn lossy_conversion(value: &JsValue, _context: &mut Context) -> JsResult<i16> {
+    match value {
+        JsValue::Rational(r) => Ok(r.round() as i16),
+        JsValue::Integer(i) => Ok(*i as i16),
+        _ => Err(JsNativeError::typ()
+            .with_message("cannot convert value to an i16")
+            .into()),
+    }
+}
+
+#[derive(Debug, TryFromJs)]
+struct TestStruct {
+    inner: bool,
+    hello: String,
+    // You can override the conversion of an attribute.
+    #[boa(from_js_with = "lossy_conversion")]
+    my_float: i16,
+}
+
+let js_str = r#"
+let x = {
+    inner: false,
+    hello: "World",
+    my_float: 2.9,
+};
+x;
+"#;
+let js = Source::from_bytes(js_str);
+
+let mut context = Context::default();
+let res = context.eval(js).unwrap();
+
+let str = TestStruct::try_from_js(&res, &mut context)
+    .unwrap();
+
```

Comment:
I took this example from the examples directory, but it seems too long. Maybe we should reduce it, or just link it, what do you think?

Reply:
I slightly reduce TryFromJs example in the latest commit, IMO it's not too big.